src/Juvix/Backends/ArithmeticCircuit/Compilation/Types.hs

```diff
@@ -1,19 +1,183 @@
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.Var var } =
+  case lookup var mem of
+    Just (n, NoExp) -> (mem, FExp $ input n)
+    Just (_, res) -> (mem, res)
+    Nothing -> panic $ show VariableOutOfScope
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.LamM{ CoreErased.body = body
+                                                          , CoreErased.arguments = arguments}} =
+  let mem' = freshVars mem arguments
+  in transTerm mem' body
+transTerm mem CoreErased.Ann{ CoreErased.term = CoreErased.AppM f params } =
+  case f of
+    CoreErased.Ann { CoreErased.term = CoreErased.LamM { CoreErased.body =  body
+                                                       , CoreErased.arguments = arguments
+                                                       }
+                   } -> let transTermParams = snd . sequenceA $ map (transTerm mem) params
+                            matchedArgParams = zip arguments transTermParams
+                            mem' = foldr (\(sy, val) m -> insert sy val m) mem matchedArgParams
+                         in transTerm mem' body
+    _ -> panic $ show SomethingWentWrongSorry
+
+transPrim :: PrimVal -> Memory -> ArithExpression Par.F
+transPrim (Element f) _ = FExp $ Lang.c f
+transPrim (Boolean b) _ = BoolExp $ Circuit.EConstBool b
+transPrim (FEInteger i) _ = FExp $ Lang.c (fromIntegral i)
+-- I miss lenses
+transPrim (BinOp Add prim prim') mem = let (_, FExp prim1) = transTerm mem prim
+                                           (_, FExp prim2) = transTerm mem prim'
+                                           in FExp $ Lang.add prim1 prim2
+transPrim (BinOp Mul prim prim') mem = let (_, FExp prim1) = transTerm mem prim
+                                           (_, FExp prim2) = transTerm mem prim'
+                                           in FExp $ Lang.mul prim1 prim2
+transPrim (BinOp Sub prim prim') mem = let (_, FExp prim1) = transTerm mem prim
+                                           (_, FExp prim2) = transTerm mem prim'
+                                           in FExp $ Lang.sub prim1 prim2
+transPrim (BinOp Eq prim prim') mem = let (_, FExp prim1) = transTerm mem prim
+                                          (_, FExp prim2) = transTerm mem prim'
+                                          in BoolExp $ Lang.eq prim1 prim2
+transPrim (BinOp And prim prim') mem = let (_, BoolExp prim1) = transTerm mem prim
+                                           (_, BoolExp prim2) = transTerm mem prim'
+                                           in BoolExp $ Lang.and_ prim1 prim2
+transPrim (BinOp Or prim prim') mem = let (_, BoolExp prim1) = transTerm mem prim
+                                          (_, BoolExp prim2) = transTerm mem prim'
+                                          in BoolExp $ Lang.or_ prim1 prim2
+-- implements exponentiation by hand
+transPrim (BinOp Exp prim CoreErased.Ann { CoreErased.term = CoreErased.Prim (FEInteger i)}) mem
```

Comment:
Looks like this will fail to pattern-match on non-constant exponentiation?

That's fine, for now, but we should explicitly error

Reply:
I think it should be caught by typechecking? You can only pass `Int` to the second parameter anyway?

Reply:
Yeah, it should be, but we should still fail gracefully